### PR TITLE
#649 Optimize memory allocation of Collider2DComponent.GetContacts

### LINKED ISSUE #649: Optimize memory allocation of Collider2DComponent.GetContacts
### Description

`Collider2DComponent.GetContacts` currently incurs dynamic memory allocation on each call if there are any contacts present for the collider. Because this API is frequently used by application/game code (e.g. query collision state every frame), the per-call allocation negatively impacts overall performance and can introduce GC pressure or frame-time spikes in real-time gameplay scenarios.

This issue proposes to optimize the memory allocation behavior of `GetContacts` (and related APIs) to minimize or eliminate allocations in the runtime path. If necessary, a low-allocation or no-allocation alternative should be exposed for use in demanding scenarios (e.g., via a `TryGetContacts`, or by introducing an overload that accepts a reusable buffer, or via an enumerable/Span-based API design).

#### Motivation
- Improve engine performance for games/applications making frequent per-frame collision queries.
- Reduce GC allocations and mitigate impact on frame-time, especially in memory-sensitive or high-performance environments.
- Follow modern .NET/C# best practices for low-allocation APIs used in tight loops.

### Relevant Code Context
- The `Collider2DComponent` type is defined in `src/Geisha.Engine/Physics/Components/Collider2DComponent.cs`.
- `GetContacts` currently returns a new collection or array of contacts, allocating memory unless no contacts are present.
- All use cases in the engine, demo apps, and test code that consume `GetContacts` are candidates for profiling and migration to the new API.
- Physics update and collision system (see `PhysicsSystem.cs`, `PreparePhysicsDebugInformation()`) may also query contacts and can benefit from reduced allocation.
- Consider adding unit/integration tests to benchmark before/after allocation behavior and verify binary compatibility of the API update.

### Acceptance Criteria
- [ ] Design and implement a low/no-allocation API for contact retrieval in `Collider2DComponent` (may be an overload, a Span/IEnumerable API, or buffer reuse pattern).
- [ ] Benchmark allocation and runtime performance with the current vs. new API in representative test scenarios.
- [ ] Update engine and demo usages to benefit from the optimized API (if possible, without breaking binary compatibility).
- [ ] Documentation: Update API comments and docs to describe allocation behavior of all `GetContacts` variants and provide migration examples.
- [ ] Unit/integration tests verify correctness and memory profile of new API (including edge cases with zero/many contacts).

## Definition of Done
- [ ] Automated Tests - new and modified code is covered with automated tests
- [ ] Documentation - new and modified APIs are documented
- [ ] Performance - performance is verified to avoid regression
- [ ] Compatibility - consider compatibility of APIs and functionality with previous version and provide upgrade path if possible
- [ ] Demo Application - if new functionality should be showcased in Demo Application or existing functionality changed then update Demo Application
- [ ] Release Activities - if release process is affected update Release Activities issue template

### PR DESCRIPTION
**Pull request by dawidkomorowski**

### CODE REVIEW

File: src/Geisha.Engine/Physics/Components/Collider2DComponent.cs [hunk 1]
```diff
@@ -60,25 +62,180 @@ protected Collider2DComponent(Entity entity) : base(entity)
...
-    ///         the result and reusing it when needed.
+    ///         <b>GC retention trade-off:</b> <see cref="Contact2D" /> is a struct that contains managed references
+    ///         (e.g. <see cref="Contact2D.ThisCollider" />, <see cref="Contact2D.OtherCollider" />). As long as the
+    ///         caller's buffer is alive and populated, those references prevent the GC from collecting the referenced
+    ///         colliders and the objects reachable from them. This is rarely a concern in scenes where entities persist
+    ///         for the lifetime of the scene, but in transition scenarios — such as reloading a level or destroying
+    ///         entities — the buffer should be cleared (e.g. <c>contacts.Clear()</c> for a <see cref="List{T}" />) to
+    ///         release those references. It is not necessary to clear the buffer on every call or every frame.
     ///     </para>
     /// </remarks>
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply src/Geisha.Engine/Physics/Components/Collider2DComponent.cs:108]: By design.

File: src/Geisha.Engine/Physics/Contact2D.cs [hunk 2]
```diff
@@ -13,10 +13,10 @@ namespace Geisha.Engine.Physics;
 ///     corresponding contact can be obtained from the perspective of <see cref="OtherCollider" /> by retrieving the
 ///     contacts from the other collider.
 /// </remarks>
-public class Contact2D
+public readonly record struct Contact2D
 {
     /// <summary>
-    ///     Initializes a new instance of the <see cref="Contact2D" /> class.
+    ///     Initializes a new instance of the <see cref="Contact2D" /> struct.
     /// </summary>
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply src/Geisha.Engine/Physics/Contact2D.cs:20]: By design.

File: test/Geisha.Engine.UnitTests/Physics/Components/RectangleColliderComponentSerializationTests.cs [hunk 3]
```diff
@@ -25,7 +25,6 @@ public void SerializeAndDeserialize()
         // Assert
         Assert.That(actual.Dimensions, Is.EqualTo(dimensions));
         Assert.That(actual.IsColliding, Is.False);
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply test/Geisha.Engine.UnitTests/Physics/Components/RectangleColliderComponentSerializationTests.cs:27]: This assertion had questionable value as the component is tested in detached state from physics system.

File: test/Geisha.Engine.UnitTests/Physics/Components/TileColliderComponentSerializationTests.cs [hunk 4]
```diff
@@ -17,7 +17,6 @@ public void SerializeAndDeserialize()
 
         // Assert
         Assert.That(actual.IsColliding, Is.False);
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply test/Geisha.Engine.UnitTests/Physics/Components/TileColliderComponentSerializationTests.cs:19]: This assertion had questionable value as the component is tested in detached state from physics system.

File: test/Geisha.Engine.UnitTests/Physics/Components/CircleColliderComponentSerializationTests.cs [hunk 5]
```diff
@@ -24,7 +24,6 @@ public void SerializeAndDeserialize()
         // Assert
         Assert.That(actual.Radius, Is.EqualTo(radius));
         Assert.That(actual.IsColliding, Is.False);
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply test/Geisha.Engine.UnitTests/Physics/Components/CircleColliderComponentSerializationTests.cs:26]: This assertion had questionable value as the component is tested in detached state from physics system.

File: test/Geisha.Engine.UnitTests/Physics/Components/Collider2DComponentTests.cs [hunk 6]
```diff
@@ -40,15 +39,4 @@ public void Constructor_ShouldCreateColliderThatIsEnabled()
         // Assert
         Assert.That(collider.Enabled, Is.True);
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply test/Geisha.Engine.UnitTests/Physics/Components/Collider2DComponentTests.cs:40]: This assertion had questionable value as the component is tested in detached state from physics system.

File: sandbox/Sandbox/engine-config.json [hunk 7]
```diff
@@ -19,7 +19,7 @@
     "ShowTotalFrames": true,
     "ShowTotalTime": true,
     //"StartUpScene": ""
-    "StartUpSceneBehavior": "AnimationSandbox"
+    "StartUpSceneBehavior": "PhysicsSandbox"
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply sandbox/Sandbox/engine-config.json:22]: Testing the physics after the optimization.

File: benchmark/Geisha.MicroBenchmark/Program.cs [hunk 8]
```diff
@@ -6,6 +6,6 @@ internal static class Program
 {
     private static void Main()
     {
-        BenchmarkRunner.Run<InputSystemBenchmarks>();
+        BenchmarkRunner.Run<GetContactsBenchmarks>();
     }
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply benchmark/Geisha.MicroBenchmark/Program.cs:10]: Currently the benchmarks are selected manually in code - exactly this line.

File: src/Geisha.Engine/Physics/Components/Collider2DComponent.cs [hunk 9]
```diff
@@ -60,25 +62,180 @@ protected Collider2DComponent(Entity entity) : base(entity)
...
+    /// </remarks>
+    public int GetContacts(List<Contact2D> contacts)
+    {
+        while (contacts.Count < ContactCount)
+        {
+            contacts.Add(default);
+        }
+
+        return GetContacts(CollectionsMarshal.AsSpan(contacts));
+    }
```
[the remark this answers is not in the document]
Comment (dawidkomorowski) [reply src/Geisha.Engine/Physics/Components/Collider2DComponent.cs:151]: Acceptable limitation. The main use case is the list provided is of sufficient capacity that it does not need to be expanded.
